includes/llms.spam.functions.php

```diff
@@ -0,0 +1,201 @@
+<?php
+/**
+ * Code related to spam detection and prevention.
+ */
+// Constants. Define these in wp-config.php to override.
+if ( ! defined( 'LLMS_SPAM_ACTION_NUM_LIMIT' ) ) {
+	define( 'LLMS_SPAM_ACTION_NUM_LIMIT', 10 );
+}
+if ( ! defined( 'LLMS_SPAM_ACTION_TIME_LIMIT' ) ) {
+	define( 'LLMS_SPAM_ACTION_TIME_LIMIT', 900 );  // in seconds
+}
+
+/**
+ * Determine whether the current visitor a spammer.
+ *
+ * @since [version]
+ *
+ * @return bool Whether the current visitor a spammer.
+ */
+function llms_is_spammer() {
+    $is_spammer = false;
+
+    $activity = llms_get_spam_activity();
+    if ( false !== $activity && count( $activity ) >= LLMS_SPAM_ACTION_NUM_LIMIT ) {
+        $is_spammer = true;
+    }
+
+	/**
+	 * Allow filtering whether the current visitor is a spammer.
+	 *
+	 * @since [version]
+	 *
+	 * @param bool  $is_spammer Whether the current visitor is a spammer.
+	 * @param array $activity   The list of potential spam activity.
+	 */
+	return apply_filters( 'llms_is_spammer', $is_spammer, $activity );
+}
+
+/**
+ * Get the list of potential spam activity.
+ *
+ * @since [version]
+ *
+ * @param string|null $ip The IP address to get activity for, or leave as null to attempt to determine current IP address.
+ *
+ * @return array|false The list of potential spam activity if successful, or false if IP could not be determined.
+ */
+function llms_get_spam_activity( $ip = null ) {
+	if ( empty( $ip ) ) {
+		$ip = llms_get_ip_address();
+	}
+
+	// If we can't determine the IP, let's bail.
+	if ( empty( $ip ) ) {
+		return false;
+	}
+
+	$ip = preg_replace( '/[^0-9a-fA-F:., ]/', '', $ip );
+	$transient_key = 'llms_spam_activity_' . $ip;
+	$activity = get_transient( $transient_key );
```

Comment:
Not sure if it's an issue using transients when they can be deleted at any time? It does allow us to use the WP system of tracking the temporary data without making our own table etc.